proposals/0000-no-kind-equalities.rst

```diff
@@ -0,0 +1,138 @@
+context: the equality is required to be satisfied in order to work with the type,
+the equality proof itself is never used. Put another way, an equality like this
+only restricts the programs you can write, never allowing new programs to be
+written.
+
+While it might be nice to eagerly forbid a type like ``T Int`` from being written,
+the benefit is slight (but real).
+
+On the other hand, the challenge in continuing to accept this syntax is real,
+and currently biting. For example, the declaration in the introduction is accepted,
+but the following is currently rejected::
+
+  type T :: forall a -> (F a ~ True) => Type
+  data T a where
+    MkT :: a -> a -> T a
+
+The reason this is rejected is that it tries to kind-check ``T a``. But because we
+don't know that ``F a ~ True``, the application ``T a`` is rejected. The "obvious"
+remedy (``MkT :: forall a -> F a ~ True => a -> a -> T a``) isn't currently available,
+as GADT syntax doesn't support visible dependent quantification (yet). Even if the
+syntax were available, it wouldn't work, because `GHC can't use an equality constraint <https://gitlab.haskell.org/ghc/ghc/-/issues/15710>`_
+in the same type signature it is introduced. We can imagine machinations that would
+accept ``T`` here, but it seems hard in the general case, most likely requiring a new
+algorithm to implement constraint generalization in kinds, a feature that is not otherwise
+needed.
+
+Even without trying to accept GADT-syntax definitions, it is challenging to
+maintain this feature. A concrete example is that Simon PJ and I (Richard)
+had to spend maybe a cumulative 8-10 hours of meetings and calls trying to
+figure out how to keep this feature working with the `work to separate
+Constraint and Type <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8750>`_.
+Instead of continuing down that path, we simply want to jettison the
+feature.
+
+It might also be interesting to note the history of this feature: I (Richard)
+added it as I was implementing ``-XTypeInType`` because it seemed easy to do
+so. No one asked for it. It just seemed like an easy addition. However, it has
+proved anything but easy. (This is not the first time we've been stymied supporting it.)
+So let's just send it back to the void whence it came.
+
+Proposed Change Specification
+-----------------------------
+
+Disallow the constraint arrow ``=>`` in the following contexts:
+
+* A kind annotation within a type: ``f ::  Int -> (Bool :: <<here>>) -> Double``
+* A visible kind application within a type: ``f :: Proxy @(<<here>>) True -> ...``
+* A kind annotation on a type variable binder: ``f :: forall (a :: <<here>>). a -> a``
+* A kind annotation on a ``data`` declaration: ``data T a b :: <<here>> where ...``
+* A kind annotation on a family declaration: ``type family F a b :: <<here>>`` (or ``data family``)
+* A standalone kind signature: ``type T :: <<here>>``
+
+In addition, any data constructor whose type includes a constraint is
```

Comment:
Does that include GADT constructors (whose signature don't look like they have constraints, but are to the best of my knowledge desugared to constructors with equality constraint parameters)?

Reply:
> Does that include GADT constructors (whose signature don't look like they have constraints, but are to the best of my knowledge desugared to constructors with equality constraint parameters)?

No, it doesn't include them.  It remains fine to have
```
data T a where
  T1 :: Int -> T Int
  ...
```
and (with `-XDataKinds`) use `T1` in a type, where it has kind, `Int -> T Int` (unsurprisingly).   But there is no equivalent, at the type level, of the GADT-aware pattern matching that we get with GADTs in terms.

Reply:
Thanks! So
```
data T a where
  T1 :: Int -> T Int
```
and
```
data T' a where
  T1' :: (Int ~ a) => Int -> T a
```
will behave different.

Reply:
Correct.  The latter can't be promoted; the former can.

We could retrench still further and prevent you promoting a GADT.  The ability to do so is not (in my understanding) particularly useful but it comes at close-to-zero cost. 

If anyone has a view here, do express it.

Reply:
I don't have a view, just questions:-)

(My personal mental model of GADTs is based only on the desugaring to the one with constraints, so it feels odd that it makes a difference, but the oddity may likely rest purely in my understanding here. And I have not used promoted GADTs yet, I think.)

Reply:
The text could make clearer that the *change* here is to forbid promotion of constructors whose constraints contain only equalities (since constraints containing classes are already unpromoteable).

Presumably the rationale for this restriction is that following this proposal it would be impossible to write the kind of a promoted constructor with an equality constraint?

How feasible would it be to continue to support promotion of such constructors anyway? I guess kind checking would need to solve wanted equality constraints arising from calls to constrained constructors, but since there would be no way to introduce given constraints, that might not be too hard? But perhaps it isn't even desirable in the absence of a way to use the evidence in (type-level) pattern matching.

Reply:
In truth, explicit equality constraints are already slightly different than GADT refinements: `MkG :: (a ~ Int) => a -> G a` has an extra field to store the equality at runtime, while `MkG :: Int -> G Int` does not. But this would make the difference more apparent, indeed.

> We could retrench still further and prevent you promoting a GADT. The ability to do so is not (in my understanding) particularly useful but it comes at close-to-zero cost.

No no no! This is one key part of `-XTypeInType`! I do not propose stopping GADT constructors from being used in types.

@adamgundry agreed about the clarification. On its way.

The rationale for the restriction is that the kind-checker doesn't need to know, at all, about constraints after this change. That is, when GHC processes a type (or a kind -- they're the same!), it must check the type for validity. This check is still completely separate from the check for terms, which comprise a different sublanguage. The validity check for types (kind-checking) will not know about constraints, a nice simplification. You surmise this in your comment; tracking constraints indeed isn't all that hard (that's why I slipped this feature in in the first place), but there are knock-on consequences, as we're now observing.

Reply:
> In truth, explicit equality constraints are already slightly different than GADT refinements: MkG :: (a ~ Int) => a -> G a has an extra field to store the equality at runtime, while MkG :: Int -> G Int does not. But this would make the difference more apparent, indeed.

Hi, @goldfirere thank you for that information. Can you please clarify, why this is the case? Is it purely for performance reasons? If equality is special cased in GADTs, why is it not special cased in ADTs using -XExistentialQuantification (but no GADT syntax)? I.e. in detail I don't understand why 

```haskell
data Dict c where
  Dict :: c => Dict
  
data Dict' c = c => Dict' 

x :: Dict (Eq Int) 
x = Dict 

x' :: Dict' (Eq Int)
x' = Dict' 
``` desugar the same while 

```haskell
type SUnit :: () -> Type
data SUnit u where SUnit :: SUnit '()

type SUnit' :: () -> Type
data SUnit' u = u ~ '() => SUnit'

y :: SUnit '()
y = SUnit

y' :: SUnit' '()
y' = SUnit'
``` desugar differently (SUnit' gets an additional `Eq#`) while pattern matching on them desugar exactly the same? 

(please excuse that this question is rather for my own interest and less relevant for the conversation)

Reply:
> Can you please clarify, why this is the case? Is it purely for performance reasons?

In GHC, equality constraints of the form `(a ~ b)` are represented as a type class. A primitive, built-in type class, but a type class nonetheless. Moreover, the return kind of `(~)` is `Constraint`, so equality constraints are represented with pointers at runtime. Unless a lot of optimizations happen, this means that equality constraints will almost certainly incur some runtime cost.

There are no equality constraints in a type like `MkG :: Int -> G Int`, so GHC is free to pick a different approach here. And indeed it does: under the hood, the type of `MkG` looks something like this:

```hs
MkG :: forall a. (a ~# Int) => Int -> G a
```

Where `(~#)` is an _unlifted_ equality. In constrast to `(~)`, which is _lifted_ (i.e., represented with a pointer at runtime), an unlifted equality is erased at runtime. This is possible because `(a ~# Int)` represents a raw coercion from `a` to `Int`, and coercions are erased when compiling from GHC's core language to lower-level intermediate languages. If you're curious about the finer details, I invite you to read [`Note [The equality types story]`](https://gitlab.haskell.org/ghc/ghc/-/blob/756a66ec0875b675bd3256d46d57419827312426/compiler/GHC/Builtin/Types/Prim.hs#L978-1136) in the GHC source code.

> If equality is special cased in GADTs, why is it not special cased in ADTs using -XExistentialQuantification (but no GADT syntax)?

The ability to write something like `MkG :: Int -> G Int` is a feature that is exclusive to GADT syntax. The closest you can get with "traditional" data constructor syntax is something like this:

```hs
data G a = (a ~ Int) => MkG Int
```

But again, that will compile differently due to the explicit, lifted equality constraint. There is no equivalent syntax for traditional data constructors.

> desugar differently (SUnit' gets an additional `Eq#`) while pattern matching on them desugar exactly the same?

I'm not sure what you mean by "pattern matching on them desugar exactly the same". That's certainly not the case when I compile this example:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
module Foo where

import Data.Kind
import Data.Type.Equality

type SUnit :: () -> Type
data SUnit u where SUnit :: SUnit '()

type SUnit' :: () -> Type
data SUnit' u = u ~ '() => SUnit'

z :: SUnit u -> u :~: '()
z SUnit = Refl

z' :: SUnit' u -> u :~: '()
z' SUnit' = Refl
```

Using GHC 9.4.4:

```
$ ghc-9.4.4 Foo.hs -fforce-recomp -ddump-simpl -dno-typeable-binds
[1 of 1] Compiling Foo              ( Foo.hs, Foo.o )

==================== Tidy Core ====================
Result size of Tidy Core
  = {terms: 20, types: 48, coercions: 9, joins: 0/0}

-- RHS size: {terms: 1, types: 1, coercions: 1, joins: 0/0}
Foo.$WSUnit [InlPrag=INLINE[final] CONLIKE] :: SUnit '()
[GblId[DataConWrapper],
 Caf=NoCafRefs,
 Unf=Unf{Src=InlineStable, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)
         Tmpl= Foo.SUnit @'() @~(<'()>_N :: '() GHC.Prim.~# '())}]
Foo.$WSUnit = Foo.SUnit @'() @~(<'()>_N :: '() GHC.Prim.~# '())

-- RHS size: {terms: 6, types: 12, coercions: 4, joins: 0/0}
z :: forall (u :: ()). SUnit u -> u :~: '()
[GblId, Arity=1, Unf=OtherCon []]
z = \ (@(u_aGC :: ())) (ds_dGY :: SUnit u_aGC) ->
      case ds_dGY of { SUnit co_aGF ->
      (Data.Type.Equality.$WRefl @() @u_aGC)
      `cast` (((:~:) <()>_N <u_aGC>_N co_aGF)_R
              :: (u_aGC :~: u_aGC) ~R# (u_aGC :~: '()))
      }

-- RHS size: {terms: 10, types: 19, coercions: 4, joins: 0/0}
z' :: forall (u :: ()). SUnit' u -> u :~: '()
[GblId, Arity=1, Unf=OtherCon []]
z'
  = \ (@(u_aGs :: ())) (ds_dGT :: SUnit' u_aGs) ->
      case ds_dGT of { SUnit' $d~_aGv ->
      case GHC.Types.eq_sel @() @u_aGs @'() $d~_aGv of co_aGM
      { __DEFAULT ->
      (Data.Type.Equality.$WRefl @() @u_aGs)
      `cast` (((:~:) <()>_N <u_aGs>_N co_aGM)_R
              :: (u_aGs :~: u_aGs) ~R# (u_aGs :~: '()))
      }
      }
```

Note that `z` matches on `SUnit` and immediately binds a coercion (i.e., unlifted equality) `co_aGF`. In `z'`, however, matching on `SUnit'` doesn't provide immediate access to a coercion. Instead, it binds `$d~_aGv`, which is a lifted equality. To get access to the underlying coercion, it has to apply the `eq_sel` function. (In GHC's Core language, `eq_sel` converts an equality class dictionary into its underlying coercion.) Applying `eq_sel` can have a runtime cost, so there is definitely a difference.

Reply:
Thank you for the long and elaborate answer, this definitely helps; the only question I'm now asking is, _why exactly_ can I not have a normal ADT witness an unlifted equality (and why does ghc not recognize that the ADT equivalent of the GADTs are isomorphic and _could_ be represented the same way)? Or am I missing something here?

Reply:
(also apparently if you don't use the coercion (in your case to construct the Refl), it doesn't put in the second case (on ghc 925), I guess that's why I assumed that it would always look the same.

Reply:
>  * We can't actually declare it that way because we don't have syntax for `~#`. And `~#` isn't a constraint, so even if we could write it, it wouldn't kind check.

(I still don't know why and why it's fundamentally impossible but I guess this explains it)

Reply:
> _why exactly_ can I not have a normal ADT witness an unlifted equality

I'm not sure how to answer this question other than to repeat what I said above: there is simply no syntax for this in the language. You could propose such a syntax, I suppose, but that would be a separate proposal.

> and why does ghc not recognize that the ADT equivalent of the GADTs are isomorphic and could be represented the same way

If we're being precise, `MkG :: forall a. (a ~ Int) => Int -> G a` and `MkG :: forall a. (a ~# Int) => Int -> G a` aren't isomorphic. This is because `(a ~# Int)` is unlifted, which means that it cannot be inhabited by bottoming values, whereas `(a ~ Int)` (a lifted type) _can_ be inhabited by bottoming values. If the compiler knows that a dictionary of type `(a ~ Int)` is non-bottoming, it is within its rights to optimize it to an unlifted equality, but it is not guaranteed that this will always happen (or even be possible).

> (I still don't know why and why it's fundamentally impossible but I guess this explains it)

Is "fundamentally impossible" referring to the inability to use `(~#)` in the source language? This is simply because `(~#)` is a magical type that isn't directly accessible. But even if you _could_ write something like `(a ~# Int) => ...` directly in the source language, that wouldn't kind-check because the thing to the left of `=>` has to be of kind `Constraint`, whereas `(#~)` has return kind `TYPE (TupleRep '[])`. (See [this dormant proposal](https://github.com/ghc-proposals/ghc-proposals/pull/116), which ran into similar issues.) Thus far, no one has created a proposal to allow this syntax that addresses these issues.

Reply:
Okay, that makes total sense, thank you. Also thank you for taking the time to explain this to me, this really helped.